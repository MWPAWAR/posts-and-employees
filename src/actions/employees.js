import fetchEmployees from '../apis/employees';
import { FETCH_EMPLOYEES, FETCH_EMPLOYEES_SUCCESS, FETCH_EMPLOYEES_FAILURE } from '../constants/action-types';

export default () => dispatch => {
  dispatch({ type: FETCH_EMPLOYEES });

  return fetchEmployees()
    .then(response => {
      dispatch({ type: FETCH_EMPLOYEES_SUCCESS, payload: { data: response.data } });
    })
    .catch(() => {
      dispatch({ type: FETCH_EMPLOYEES_FAILURE });
    })
};
