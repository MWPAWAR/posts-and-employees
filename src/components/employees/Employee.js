import React from 'react';
import PropTypes from 'prop-types';
import Card from '@material-ui/core/Card';
import CardActionArea from '@material-ui/core/CardActionArea';
import CardActions from '@material-ui/core/CardActions';
import CardContent from '@material-ui/core/CardContent';
import Button from '@material-ui/core/Button';
import Typography from '@material-ui/core/Typography';
import { Link } from 'react-router-dom';
import { FormattedMessage } from 'react-intl';

const Employee = ({
  id,
  address: { suite, street, city, zipcode },
  name,
}) => (
  <Card className="root">
    <CardActionArea>
      <CardContent>
        <Typography gutterBottom variant="h5" component="h2">{name}</Typography>
        <Typography variant="body1" color="textSecondary" component="p">
          <FormattedMessage
            id="app.employees.address"
            values={{ suite, street, city, zipcode }}
          />
        </Typography>
      </CardContent>
    </CardActionArea>
    <CardActions>
      <Link to={`/employees/${id}`}>
        <Button size="large" color="primary">
          <FormattedMessage id="app.employees.btn-view-label" />
        </Button>
      </Link>
    </CardActions>
  </Card>
);

Employee.propTypes = {
  id: PropTypes.number.isRequired,
  name: PropTypes.string.isRequired,
  address: PropTypes.shape({
    street: PropTypes.string.isRequired,
    city: PropTypes.string.isRequired,
    suite: PropTypes.string.isRequired,
    zipcode: PropTypes.string.isRequired
  }).isRequired
};

export default Employee;
