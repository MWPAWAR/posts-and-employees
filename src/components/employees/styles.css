.heading {
  margin: 1rem 0;
  text-align: center;
}

.container {
  display: flex;
  flex-wrap: wrap;
}

.employee-details {
  margin: 1rem;
}
