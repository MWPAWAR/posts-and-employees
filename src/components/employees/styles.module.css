.heading {
  margin: 1rem 0;
  text-align: center;
}

.container {
  display: flex;
  flex-wrap: wrap;
}

.employeeDetails {
  margin: 1rem;
  width: 100%;
}
