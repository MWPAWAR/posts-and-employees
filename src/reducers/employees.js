import { FETCH_EMPLOYEES, FETCH_EMPLOYEES_SUCCESS, FETCH_EMPLOYEES_FAILURE } from '../constants/action-types';

const INITIAL_STATE = { isError: false };

export default (state = INITIAL_STATE, action) => {
  const { type, payload } = action;

  switch(type) {
    case FETCH_EMPLOYEES:
      return { ...state };
    case FETCH_EMPLOYEES_SUCCESS:
      return { ...state, ...payload, isError: false };
    case FETCH_EMPLOYEES_FAILURE:
      return { ...state, isError: true };
    default:
      return state;
  }
};
