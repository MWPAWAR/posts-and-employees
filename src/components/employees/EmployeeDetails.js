import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { FormattedMessage } from 'react-intl';

import Card from '../core/card/Card';
import DetailsRow from '../core/details-row/DetailsRow';
import Error from '../core/error/Error';
import getEmployeeById from '../../selectors/employee';
import './styles.css';

const EmployeeDetails = ({ employee }) => {
  if (!employee) {
    return (
      <Error
        message={<FormattedMessage id="app.error-no-data" />}
      />
    )
  }

  const {
    name,
    username,
    email,
    phone,
    website,
    address,
    company
  } = employee;

  return (
    <Card className="employee-details">
      <h3>
        <FormattedMessage
          id="app.employees.employee.name"
          values={{ name, username }}
        />
      </h3>
      <DetailsRow
        label={<FormattedMessage id="app.employees.employee.email" />}
        value={email}
      />
      <DetailsRow
        label={<FormattedMessage id="app.employees.employee.phone" />}
        value={phone}
      />
      <DetailsRow
        label={<FormattedMessage id="app.employees.employee.website" />}
        value={<a href={`//${website}`} target="blank">{website}</a>}
      />
      <DetailsRow
        label={<FormattedMessage id="app.employees.employee.address-label" />}
        value={
          <FormattedMessage
            id="app.employees.employee.address-val"
            values={{ suite: address.suite, street: address.street, city: address.city, zipcode: address.zipcode }}
          />
        }
      />
      <DetailsRow
        label={<FormattedMessage id="app.employees.employee.company" />}
        value={
          <FormattedMessage
            id="app.employees.employee.company-val"
            values={{ companyName: company.name, catchPhrase: company.catchPhrase }}
          />
        }
      />
    </Card>
  );
};

const mapStateToPros = (state, props) => ({
  employee: getEmployeeById(state, props.match.params.empId)
});

EmployeeDetails.propTypes = {
  employee: PropTypes.shape({
    name: PropTypes.string.isRequired,
    username: PropTypes.string.isRequired,
    email: PropTypes.string.isRequired,
    phone: PropTypes.string.isRequired,
    website: PropTypes.string.isRequired,
    address: PropTypes.shape({
      street: PropTypes.string.isRequired,
      suite: PropTypes.string.isRequired,
      city: PropTypes.string.isRequired,
      zipcode: PropTypes.string.isRequired,
    }),
    company: PropTypes.shape({
      name: PropTypes.string.isRequired,
      catchPhrase: PropTypes.string.isRequired,
      bs: PropTypes.string.isRequired,
    })
  })
};

export { EmployeeDetails };

export default connect(mapStateToPros)(EmployeeDetails);
