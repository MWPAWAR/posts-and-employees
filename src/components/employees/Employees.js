import React, { useEffect, Fragment } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { FormattedMessage } from 'react-intl';

import Employee from './Employee';
import getEmployees from '../../selectors/employees';
import fetchEmployees from '../../actions/employees';
import './styles.css'

const Employees = ({
  employees,
  fetchEmployees,
  hasEmployees
}) => {
  useEffect(() => {
    if (!hasEmployees) fetchEmployees();
  }, []);

  const employeeItems = 
    employees.map((employee) => (
      <Employee
        id={employee.id}
        name={employee.name}
        address={employee.address}
      />
    )
  );

  return (
    <Fragment>
      <h2 className="heading">
        <FormattedMessage id="app.employees.label" />
      </h2>
      <div className="container">{employeeItems}</div>
    </Fragment>
  );
};

const mapStateToProps = state => {
  const employees = getEmployees(state);
  const hasEmployees = employees.length > 0;

  return {
    employees,
    hasEmployees
  };
};

Employees.propTypes = {
  employees: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      name: PropTypes.string.isRequired,
      address: PropTypes.object.isRequired
    })
  ),
  fetchEmployees: PropTypes.func.isRequired,
  hasEmployees: PropTypes.bool.isRequired
};

export { Employees }; // exporting a constant from here for testing.

export default connect(mapStateToProps, { fetchEmployees })(Employees);
