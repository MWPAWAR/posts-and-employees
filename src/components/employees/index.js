import Employee from './Employee';
import EmployeeDetails from './EmployeeDetails';
import Employees from './Employees';

export { Employee, EmployeeDetails, Employees };
