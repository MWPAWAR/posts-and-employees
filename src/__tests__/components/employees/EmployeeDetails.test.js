import React from 'react';
import { shallow } from 'enzyme';

import { EmployeeDetails } from '../../../components/employees/EmployeeDetails';
import employeeMock from '../../mocks/employee.json';

describe('EmployeeDetails', () => {
  test('renders data unavailable message if employee is undefined', () => {
    const component = shallow(<EmployeeDetails />);
    expect(component).toMatchSnapshot();
  });

  test('renders correctly with default data', () => {
    const component = shallow(<EmployeeDetails employee={employeeMock} />);
    expect(component).toMatchSnapshot();
  });
});
