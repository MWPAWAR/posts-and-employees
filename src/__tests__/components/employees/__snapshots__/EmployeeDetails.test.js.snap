// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmployeeDetails renders correctly with default data 1`] = `
<Card
  className="employeeDetails"
>
  <h3>
    <FormattedMessage
      id="app.employees.employee.name"
      values={
        Object {
          "name": "Glenna Reichert",
          "username": "Delphine",
        }
      }
    />
  </h3>
  <DetailsRow
    label={
      <FormattedMessage
        id="app.employees.employee.email"
      />
    }
    value="Chaim_McDermott@dana.io"
  />
  <DetailsRow
    label={
      <FormattedMessage
        id="app.employees.employee.phone"
      />
    }
    value="(775)976-6794 x41206"
  />
  <DetailsRow
    label={
      <FormattedMessage
        id="app.employees.employee.website"
      />
    }
    value={
      <a
        href="//conrad.com"
        rel="noopener noreferrer"
        target="_blank"
      >
        conrad.com
      </a>
    }
  />
  <DetailsRow
    label={
      <FormattedMessage
        id="app.employees.employee.address-label"
      />
    }
    value={
      <FormattedMessage
        id="app.employees.employee.address-val"
        values={
          Object {
            "city": "Bartholomebury",
            "street": "Dayna Park",
            "suite": "Suite 449",
            "zipcode": "76495-3109",
          }
        }
      />
    }
  />
  <DetailsRow
    label={
      <FormattedMessage
        id="app.employees.employee.company"
      />
    }
    value={
      <FormattedMessage
        id="app.employees.employee.company-val"
        values={
          Object {
            "catchPhrase": "Switchable contextually-based project",
            "companyName": "Yost and Sons",
          }
        }
      />
    }
  />
</Card>
`;

exports[`EmployeeDetails renders data unavailable message if employee is undefined 1`] = `
<DataUnavailable
  message={
    <FormattedMessage
      id="app.error-no-data"
    />
  }
/>
`;
